Enable cloud-init and system networking for hyperv provider

lstocchi opened the pull request:
This PR has been built over https://github.com/cfergeau/podman/pull/8
It then adds the missing pieces to have a working workflow with macadam/hyperv (podman should work fine as before).

This contains several commits that do:
1. make ignition optional, so that if cloud-init is true, all ignition code is skipped
2. add code to generate the cloud-init config ISO - it will then be attached as a DVD drive to the VM
3. when using system networking, it retrieves the VM IP address and uses the default SSH port to connect to it

anjannath committed:
hyperv: enable to run VM without gvproxy

So far, hyperv only worked with usermode networking enabled, even if the
user did not explicitly use the --user-mode-networking flag. However,
macadam needs a way to disable usermode networking on hyperv.

Because of this, the UseProviderNetworkSetup func has been updated to accept
MachineConfig as param so its return a value calculated using the actual
userModeNetworking value.

The UserModeNetworking flag has been added to
HyperVConfig so it can be persisted when saving the MachineConfig.

If system networking is enabled it does not create a hvsock for the
networking.

lstocchi committed:
fix Groups type field in CloudInit User

The groups field of a cloudInit user could have a list of string as a
value. This patch update the field type

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
skip ignition setup if cloud-init is enabled

this commit skips all code related to ignition when cloud-init is in use.
It affects all providers, mainly applehv, libkrun and hyperv.
For hyperv it also adds checks when removing hvsock entries
in the Registry.

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
generate cloud-init ISO for hyperv

it generate an ISO including cloud-init config to be attached to the VM

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
when using system networking it should retrieve IP and use default SSH port

if system networking mode is enabled it tries to detect the VM's IP after
it get started and change the SSH port to its default value 22.
This comes in handy when trying to check if the VM is running or to SSH
in it.

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
extend ssh check to use ip address

it uses the VM IP and default SSH port to check if the VM is running and
connect to it

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
upgrade libhvee dep

it upgrades the libhvee library version. To use cloud-init on Hyperv we need to
attach a DVD drive to the VM so we needs a version of libhvee that
includes https://github.com/containers/libhvee/pull/235

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
set user mode networking based on cloud init value on hyperv

this commit sets the hyperv user mode networking value based if cloud-init is
used or not. This way we maintain compatibility with podman.

So far, hyperv only worked with usermode networking enabled, even if the
user did not explicitly use the --user-mode-networking flag. However,
macadam needs a way to disable usermode networking on hyperv.

To maintain the current podman behavior and still provide a way to
macadam to disable usermode we use cloud-init feature. Only macadam
relies on it, so if cloud-init is enabled, we use system networking.

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi commented:
You can use https://github.com/crc-org/macadam/pull/193 for testing this with macadam

cfergeau commented on go.sum:
```diff
@@ -76,8 +76,8 @@ github.com/containers/gvisor-tap-vsock v0.8.6 h1:9SeAXK+K2o36CtrgYk6zRXbU3zrayjv
-github.com/containers/libhvee v0.10.0 h1:7VLv8keWZpHuGmWvyY4c1mVH5V1JYb1G78VC+8AlrM0=
-github.com/containers/libhvee v0.10.0/go.mod h1:at0h8lRcK5jCKfQgU/e6Io0Mw12F36zRLjXVOXRoDTM=
+github.com/containers/libhvee v0.10.1-0.20250623125428-422aa7ddc0e5 h1:wRVaP8192oHKkQ7hHLm+Dsq+08Mt4GKl4SDF/jhm5Cs=
+github.com/containers/libhvee v0.10.1-0.20250623125428-422aa7ddc0e5/go.mod h1:MLfTB8sW7E2UwQ/WknTHbuiYtNpWCR9XENMhA/hm8h0=
```
The `go mod vendor` changes are missing in this commit

cfergeau commented on pkg/machine/shim/host.go:
```diff
@@ -280,12 +285,12 @@ func Init(opts machineDefine.InitOptions, mp vmconfigs.VMProvider) error {
 	}
 
-	err = mp.CreateVM(createOpts, mc, &ignBuilder)
+	err = mp.CreateVM(createOpts, mc, ignBuilder)
```
Any reason for not passing a `nil` `ignBuilder` here when `opts.CloudInit` is set?

lstocchi replied:
replied in the commit below which cover the same stuff. Let me know if you prefer to not using a pointer and have another if/else here

cfergeau replied:
It's fine, I missed that this was already handled :)

cfergeau commented on pkg/machine/shim/networking.go:
```diff
@@ -145,7 +145,11 @@ func conductVMReadinessCheck(mc *vmconfigs.MachineConfig, maxBackoffs int, backo
 			continue
 		}
-		if !isListening(mc.SSH.Port) {
+		address := "127.0.0.1"
```
The rest of the code uses "localhost" for this.

cfergeau commented on pkg/machine/ssh.go:
```diff
@@ -17,28 +18,40 @@ import (
+}
+
+func CommonSSHVMSilent(mc *vmconfigs.MachineConfig, inputArgs []string) error {
+	address := mc.HyperVHypervisor.IPAddress
```
Does this work as expected on non-Windows builds? unexpected to see `HyperV` in this generic code.
I think I’d add a `CommonSSHSilentWithAddress` or such with an `address` arg.

cfergeau commented on pkg/machine/ssh.go:
```diff
@@ -17,28 +18,40 @@ import (
+	return commonBuiltinSSH(username, identityPath, name, "localhost", sshPort, inputArgs, true, os.Stdin)
+}
+
+func CommonSSHShellWithAddress(username, identityPath, name, address string, sshPort int, inputArgs []string) error {
```
Can this be private to the ssh module? (ie named `commonSSHShell…` instead of `CommonSSHShell…`?)

lstocchi replied:
This should be called from the ssh command in macadam. This is why it's not private

cfergeau commented on pkg/machine/shim/host.go:
```diff
@@ -195,74 +190,84 @@ func Init(opts machineDefine.InitOptions, mp vmconfigs.VMProvider) error {
-		err = ignBuilder.BuildWithIgnitionFile(opts.IgnitionPath)
-
+	playbookDest := fmt.Sprintf("/home/%s/%s", userName, "playbook.yaml")
+	var playbookContents string
```
To minimize the amount of changes, and keep `playbookDest` and `playbookContents` inside the `if len(opts.PlaybookPath) > 0` block, I think I’d change `err = ignBuilder.AddPlaybook(playbookContents, playbookDest, userName)` to `err = ignBuilder.AddPlaybook(mc.Ansible.Contents, mc.Ansible.PlaybookPath, m, userName)` in a commit before this one.

cfergeau commented on pkg/machine/shim/host.go:
```diff
@@ -195,74 +190,84 @@ func Init(opts machineDefine.InitOptions, mp vmconfigs.VMProvider) error {
+			WritePath: ignitionFile.GetPath(),
+			Rootful:   opts.Rootful,
+		})
+		ignBuilder = &tmpIgnBuilder
```
Not fully sure why a `*IgnitionBuilder` is needed instead of an `IgnitionBuilder`? In any case, would `ignBuilder := &ignition.NewIgnitionBuilder(ignition.DynamicIgnition{…` work to avoid `ignBuilder`?

lstocchi replied:
I was using a pointer so if it was not initialized it had a nil value assigned by default and i did not have to add an if/else to send a nil/not nil value when calling CreateVM `err = mp.CreateVM(createOpts, mc, ignBuilder)`

The fact that i used a tmp variable is that if i don't use it i face `invalid operation: cannot take address of ignition.NewIgnitionBuilder(ignition.DynamicIgnition{…}) (value of struct type ignition.IgnitionBuilder)`
when using 

```
ignBuilder = &ignition.NewIgnitionBuilder(ignition.DynamicIgnition{.......
```

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -216,15 +221,21 @@ func (h HyperVStubber) StartVM(mc *vmconfigs.MachineConfig) (func() error, func(
+		waitReady, listener, err = mc.HyperVHypervisor.ReadyVsock.ListenSetupWait()
+		if err != nil {
+			return nil, nil, err
+		}
```
This commit is already fairly large and at times hard to follow, it would be nice to split in its own commit the part related to "For hyperv it also adds checks when removing hvsock entries in the Registry."

lstocchi replied:
Done 👍

cfergeau reviewed (commented):
Need to look more closely at the "usermode networking" part

lstocchi committed:
check and remove hvsock entries in the Registry

when using cloudInit podman does not add any hvsock to the Registry.
This patch adds some check to prevent deleting unexisting Registry
entries

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi commented:
@cfergeau i should have fixed/replied to all your comments. Feel free to check it again. Thanks!! 👍

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -493,6 +507,44 @@ func resizeDisk(newSize strongunits.GiB, imagePath *define.VMFile) error {
+}
+
+func getIPAddress(name string) (string, error) {
+	ipAddress := exec.Command("powershell", []string{"-command", fmt.Sprintf("Get-VM -Name %s | Select-Object -ExpandProperty NetworkAdapters | Select-Object IPAddresses", name)}...)
```
Do you know if there's an API for this in libhvee? I could not find anything when I looked but maybe I missed it.

lstocchi replied:
No i did not find anything, but i guess they did not need anything similar as they were using usermode networking

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -28,15 +28,14 @@ import (
-func (h HyperVStubber) UseProviderNetworkSetup() bool {
-	return h.UserModeNetworking == false
+func (h HyperVStubber) UseProviderNetworkSetup(mc *vmconfigs.MachineConfig) bool {
+	return mc.HyperVHypervisor.UserModeNetworking == false
```
wondering if this should be squashed with the first commit which is adding this?

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -58,6 +57,14 @@ func (h HyperVStubber) CreateVM(opts define.CreateVMOpts, mc *vmconfigs.MachineC
+		mc.HyperVHypervisor = new(vmconfigs.HyperVConfig)
+	}
+
+	// Set userModeNetworking based on cloudInit value for backwards compatibility
```
I would mention that usermode networking with hyperv requires gvforwarder in the guest, and the cloud init code cannot inject it for now, so it has to be disabled.

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -58,6 +57,14 @@ func (h HyperVStubber) CreateVM(opts define.CreateVMOpts, mc *vmconfigs.MachineC
+	}
+
+	// Set userModeNetworking based on cloudInit value for backwards compatibility
+	mc.HyperVHypervisor.UserModeNetworking = !mc.CloudInit
```
Does this have to be in `mc.HyperVHypervisor`? Could this be `h.UserModeNetworking = !mc.CloudInit`?

cfergeau replied:
Below, some comment I wrote during the review, but I'm really not sure it's relevant. Let's ignore this comment for now, I'm only recording it so that we can get back to it later. In general, I find it hard to track where/how the default value for user mode networking is set

Just food for thoughts, I'm not asking to solve this in this PR:

> I'm having second thoughts about this approach. Seeing how the wsl code is structured:
https://github.com/containers/podman/blob/823487970876ee48b7b32b6fc7fd255b4b4c73ef/pkg/machine/wsl/stubber.go#L39-L44
>
> and seeing this in `InitOpts` https://github.com/containers/podman/blob/823487970876ee48b7b32b6fc7fd255b4b4c73ef/pkg/machine/define/initopts.go#L22
>
> maybe we could do this?
> - set `UserModeNetworking` to `false` when macadam create `InitOpts` on hyperv
> - error out in `hyperv/stubber.go` if `UserModeNetworking` is true and cloud-init is in use
(not sure this could work with

lstocchi replied:
> Does this have to be in `mc.HyperVHypervisor`? Could this be `h.UserModeNetworking = !mc.CloudInit`?

No bc h.UserModeNetworking will not have the updated value. The stubber is just initialized when the command is executed `new(stubber)` so we must rely on mc to have the actual value.

cfergeau commented on pkg/machine/ssh.go:
```diff
@@ -17,28 +18,36 @@ import (
+	return commonBuiltinSSH(username, identityPath, name, "localhost", sshPort, inputArgs, true, stdin)
+}
+
+func CommonSSHVMSilent(mc *vmconfigs.MachineConfig, address string, inputArgs []string) error {
```
The rest of the functions in this module seem to avoid using `github.com/containers/podman/v5/pkg/machine/vmconfigs` and to use a full list of args, I'd favor not adding the dependency and also have 
`func CommonSSHSilentWithAddress(username, identityPath, name, address string, sshPort int, inputArgs []string)`

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -469,17 +478,23 @@ func resizeDisk(newSize strongunits.GiB, imagePath *define.VMFile) error {
-	if err := mc.HyperVHypervisor.NetworkVSock.Remove(); err != nil {
-		logrus.Errorf("unable to remove registry entry for %s: %q", mc.HyperVHypervisor.NetworkVSock.KeyName, err)
-	}
+	if mc.HyperVHypervisor != nil {
```
Can this be `nil`? it's now initialized very early in `CreateVM`

cfergeau commented on pkg/machine/hyperv/stubber.go:
```diff
@@ -218,15 +221,21 @@ func (h HyperVStubber) StartVM(mc *vmconfigs.MachineConfig) (func() error, func(
-		return nil, nil, err
+	var waitReady func() error
+	var listener io.Closer
+	if mc.HyperVHypervisor != nil && mc.HyperVHypervisor.ReadyVsock.KeyName != "" {
```
Can `mc.HyperVHypervisor` be `nil`?
